corehq/apps/api/user_updates.py

```diff
@@ -165,3 +167,53 @@ def _log_phone_number_change(new_phone_numbers, old_phone_numbers, user_change_l
 
     if change_messages:
         user_change_logger.add_change_message({'phone_numbers': change_messages})
+
+
+def _update_location(user, location_object, user_change_logger):
+    primary_location_id = location_object.get('primary_location')
+    location_ids = location_object.get('locations')
+
+    if primary_location_id is None and location_ids is None:
+        return
+
+    user_current_primary_location_id = user.location_id
+
+    if not primary_location_id and not location_ids:
+        # we will have bundle.obj.save() later, so when change locations, commit=False
+        user.unset_location(commit=False)
+        user.reset_locations([], commit=False)
+        if user_change_logger:
+            user_change_logger.add_changes({'location_id': primary_location_id})
+            user_change_logger.add_info(UserChangeMessage.primary_location_removed())
+            user_change_logger.add_changes({'assigned_location_ids': location_ids})
+            user_change_logger.add_info(UserChangeMessage.assigned_locations_info([]))
+    elif not primary_location_id or not location_ids:
+        raise UpdateUserException(_('Both primary_location and locations must be provided together.'))
+    else:
+        if primary_location_id not in location_ids:
+            raise UpdateUserException(_('Primary location must be included in the list of locations.'))
+
+        # Verify all location ids has a SQLLocation.
+        for loc_id in location_ids:
+            try:
+                SQLLocation.active_objects.get(location_id=loc_id, domain=user.domain)
+            except SQLLocation.DoesNotExist:
+                raise UpdateUserException(f"Could not find location id '{loc_id}'.")
+
+        if primary_location_id != user_current_primary_location_id:
+            primary_location = SQLLocation.active_objects.get(location_id=primary_location_id)
+            user.set_location(primary_location, commit=False)
```

Comment:
`set_location` also takes a location ID, so you can skip the lookup.  Or is it necessary for the logging?  If you do need the location objects anyways, then I'd just fetch them all a couple lines above when you check that they're real, to avoid extra DB hits.

Reply:
It is necessary for the logging. I just realized I query all the locations twice, so I updated in 106745fccd65bba25d3a47c9be960b01ae8e362c. But I don't know how to get the `primary_location` object from that query? The only thing I can think of is iterate through all the location objects, then compare the location id and `primary_location_id`